python/tests/test_tables.py

```diff
@@ -2529,3 +2529,75 @@ def test_empty_nodes(self):
             self.assertEqual(subset.provenances, tables.provenances)
+
+
+class TestUnion(unittest.TestCase):
+    """
+    Tests for the TableCollection.union method
+    """
+
+    def get_msprime_example(self, sample_size=3, T=5, seed=1239):
+        # we assume after the split the ts are completely independent
+        M = [[0, 0], [0, 0]]
+        population_configurations = [
+            msprime.PopulationConfiguration(sample_size=sample_size),
+            msprime.PopulationConfiguration(sample_size=sample_size),
+        ]
+        demographic_events = [
+            msprime.CensusEvent(time=T),
+            msprime.MassMigration(T, source=1, dest=0, proportion=1),
+        ]
+        ts = msprime.simulate(
+            population_configurations=population_configurations,
+            demographic_events=demographic_events,
+            migration_matrix=M,
+            length=2e5,
+            recombination_rate=1e-8,
+            mutation_rate=1e-7,
+            record_migrations=False,
+            random_seed=seed,
+        )
+        ts = tsutil.add_random_metadata(ts, seed)
+        ts = tsutil.insert_random_ploidy_individuals(ts, max_ploidy=1)
+        return ts
+
+    def get_wf_example(self, N=5, T=5, seed=1249):
+        twopop_tables = wf.wf_sim(N, T, num_pops=2, seed=seed, deep_history=True)
+        twopop_tables.sort()
+        ts = twopop_tables.tree_sequence()
+        ts = ts.simplify()
+        # adding muts
+        ts = tsutil.jukes_cantor(ts, 1, 10, seed=seed)
+        ts = tsutil.add_random_metadata(ts, seed)
+        ts = tsutil.insert_random_ploidy_individuals(ts, max_ploidy=2)
+        return ts
+
+    def split_example(self, ts, T):
+        # splitting two pop ts into disjoint ts
+        shared_nodes = [n.id for n in ts.nodes() if n.time >= T]
+        pop1 = list(ts.samples(population=0))
+        pop2 = list(ts.samples(population=1))
+        tables1 = ts.simplify(shared_nodes + pop1, record_provenance=False).tables
+        tables2 = ts.simplify(shared_nodes + pop2, record_provenance=False).tables
+        nodes = [
+            i if i < len(shared_nodes) else tskit.NULL
+            for i in range(tables2.nodes.num_rows)
+        ]
+        return tables1, tables2, nodes
+
+    def get_examples(self, N=5, T=5, seed=242):
+        yield self.split_example(self.get_msprime_example(N, T, seed=seed), T)
+        yield self.split_example(self.get_wf_example(N, T, seed=seed), T)
+
+    def verify_union_equality(self, tables, other, nodes):
+        uni1 = tables.copy()
+        uni2 = tables.copy()
+        uni1.union(other, nodes, record_provenance=False)
+        tsutil.py_union(uni2, other, nodes, record_provenance=False)
+        self.assertEqual(uni1, uni2)
+
+    def test_msprime_examples(self):
+        for N in [2, 4, 5]:
+            for T in [2, 5, 20]:
+                for example in self.get_examples(N, T, 9257):
+                    self.verify_union_equality(*example)
```

Comment:
We should think of a couple of other very simple tests that don't rely on just comparing to the python implementation. how about: (a) if there are no shared nodes then one should just be stacked on top the other; (b) if all the nodes are shared then it shouldn't change self; (c) in verify_union_equality, simplifying to the original set of nodes should do just the same thing before and after the union?

Reply:
Also, we should test the errors here: (a) the node map not having the right number of entries, and (b) the node map referring to nonsensical nodes: eg wrong type (floats) or an integer that is self.num_nodes or greater.

Reply:
1(a) is not that simple given we implictly sort stuff, but the other suggestions are good! thanks!

Reply:
It would be good to check the 'no shared nodes' case somehow...

Reply:
that's right, we have no cases for that right now. well, we could check that case against the py version anyway.

Reply:
I think I am covering the errors in `tests_lowlevel.py`. here:

https://github.com/tskit-dev/tskit/blob/56e573ecf464c3d852e208f40c4dfc0eba3a704c/python/tests/test_lowlevel.py#L280-L293